File: WiertarBot/commands/special.py
from typing import cast

import fbchat

from ..message_dispatch import MessageEventDispatcher
from ..events import Mention, MessageEvent
from ..response import response
from ..services import PermissionService


@MessageEventDispatcher.register(special=True)
async def everyone(event: MessageEvent):
    if '@everyone' in event.getText() \
            and PermissionService.isAuthorized('everyone', event.getThreadId(), event.getAuthorId()) \
            and event.isGroup():
        group = cast(fbchat.GroupData, await event.getContext().fetch_thread(event.getThreadId()))
        mentions = [
            Mention(thread_id=participant.id, offset=0, length=9)
            for participant in group.participants
        ]

        await response(event, text="@everyone", mentions=mentions).pySend()


@MessageEventDispatcher.register(special=True)
async def thinking(event: MessageEvent):
    if event.getText() == '🤔':
        await response(event, text='🤔').pySend()


@MessageEventDispatcher.register(special=True)
async def grek(event: MessageEvent):
    text = event.getText().lower()
    if text == 'grek':
        if event.getText() == 'Grek':
            await response(event, text="grek*").pySend()
        await response(event, text="to pedał").pySend()
    elif text == 'pedał':
        await response(event, text="sam jesteś grek").pySend()
    elif text == 'pedał to':
        await response(event, text="grek").pySend()


@MessageEventDispatcher.register(special=True)
async def leet(event: MessageEvent):
    if '1337' in event.getText():
        p = PermissionService.isAuthorized('leet', event.getThreadId(), event.getAuthorId())

        await response(event, text="Jesteś elitą" if p else "Nie jesteś elitą").pySend()


@MessageEventDispatcher.register(special=True)
async def papiezowa_liczba(event: MessageEvent):
    if '2137' in event.getText():
        await response(event, text='haha toż to papieżowa liczba').pySend()


@MessageEventDispatcher.register(special=True)
async def Xd_reaction(event: MessageEvent):
    if 'Xd' in event.getText():
        await event.pyReact('😠')  # angry reaction


async def spierwyp(event: MessageEvent, word: str):
    text = event.getText().lower()
    msg = 'sam '

    if text.startswith('sam') and text.endswith(word):
        t = text.replace(' ', '').replace('sam', '').replace(word, '')
        if t == '' and text.count(word) == 1:
            msg = 'sam ' * (text.count('sam') + 1)

    if word in text:
        await response(event, text=msg + word).pySend()
        await event.pyReact('😠')  # angry reaction


@MessageEventDispatcher.register(special=True)
async def spier(event: MessageEvent):
    await spierwyp(event, 'spierdalaj')


@MessageEventDispatcher.register(special=True)
async def wyp(event: MessageEvent):
    await spierwyp(event, 'wypierdalaj')

File: services/core/src/main/kotlin/pl/kvgx12/wiertarbot/commands/Special.kt
package pl.kvgx12.wiertarbot.commands

import kotlinx.coroutines.coroutineScope
import kotlinx.coroutines.launch
import org.springframework.beans.factory.BeanRegistrarDsl
import pl.kvgx12.wiertarbot.command.dsl.CommandDsl
import pl.kvgx12.wiertarbot.command.dsl.command
import pl.kvgx12.wiertarbot.command.dsl.special
import pl.kvgx12.wiertarbot.command.dsl.specialCommandName
import pl.kvgx12.wiertarbot.proto.MessageEvent
import pl.kvgx12.wiertarbot.proto.mention
import pl.kvgx12.wiertarbot.services.PermissionService
import pl.kvgx12.wiertarbot.utils.proto.Response
import pl.kvgx12.wiertarbot.utils.proto.isGroup

const val THINKING_EMOJI = "\uD83E\uDD14"
const val ANGRY_EMOJI = "\uD83D\uDE20"

private val String.special get() = specialCommandName(this)

class SpecialCommandsRegistrar : BeanRegistrarDsl({
    command("everyone".special) {
        val permissionService = dsl.bean<PermissionService>()

        special { event ->
            if ("@everyone" in event.text &&
                event.isGroup &&
                permissionService.isAuthorized("everyone", event.threadId, event.authorId)
            ) {
                val mentions = event.context.fetchThread(event.threadId)!!
                    .participantsList.map {
                        mention {
                            threadId = it.id
                            offset = 0
                            length = 9
                        }
                    }

                Response(event, text = "@everyone", mentions = mentions).send()
            }
        }
    }

    command("thinking".special) {
        special {
            if (it.text == THINKING_EMOJI) {
                Response(it, text = THINKING_EMOJI).send()
            }
        }
    }

    command("grek".special) {
        special {
            when (it.text.lowercase()) {
                "grek" -> {
                    if (it.text == "Grek") {
                        Response(it, text = "grek*").send()
                    }
                    Response(it, text = "to pedał").send()
                }

                "pedał" -> Response(it, text = "sam jesteś grek").send()
                "pedał to" -> Response(it, text = "grek").send()
            }
        }
    }

    command("1337".special) {
        val permissionService = dsl.bean<PermissionService>()

        special {
            if ("1337" in it.text) {
                val isLeet = permissionService.isAuthorized("leet", it.threadId, it.authorId)

                Response(it, text = if (isLeet) "Jesteś elitą" else "Nie jesteś elitą").send()
            }
        }
    }

    command("2137".special) {
        special {
            if ("2137" in it.text) {
                Response(it, text = "haha toż to papieżowa liczba").send()
            }
        }
    }

    command("Xd".special) {
        special {
            if ("Xd" in it.text) it.context.reactToMessage(it, ANGRY_EMOJI)
        }
    }

    command("spierdalaj".special) {
        special {
            sam(it, "spierdalaj")
        }
    }

    command("wypierdalaj".special) {
        special {
            sam(it, "wypierdalaj")
        }
    }
})

private suspend fun CommandDsl.sam(event: MessageEvent, word: String) = coroutineScope {
    val text = event.text.lowercase()

    if (word !in text) return@coroutineScope

    var msg = "sam "

    if (text.startsWith("sam") && text.endsWith(word)) {
        val t = text.replace(" ", "")
            .replace("sam", "")
            .replaceFirst(word, "")

        if (t.isEmpty()) {
            msg = "sam ".repeat(text.split("sam").size)
        }
    }

    launch { Response(event, text = msg + word).send() }
    launch { event.context.reactToMessage(event, ANGRY_EMOJI) }
}
